xbmc/utils/Mp4ChplReader.h

```diff
@@ -0,0 +1,106 @@
+
+class CURL;
+namespace XFILE
+{
+class CFile;
+}
+
+struct ChplChapter
+{
+  uint64_t start_pts_ms = 0; // milliseconds
+  std::string title;
+};
+
+enum class ChplChapterStatus : int
+{
+  Error = -1,
+  Found = 0,
+  None = 1
+};
+
+constexpr auto chplError = ChplChapterStatus::Error;
+constexpr auto chplFound = ChplChapterStatus::Found;
+constexpr auto chplNone = ChplChapterStatus::None;
+
+struct ChplChapterResult
+{
+  ChplChapterStatus status{ChplChapterStatus::None};
+  std::optional<int64_t> value; // For atom size
+  std::optional<std::string> errorMessage; // Optional error description
+
+  constexpr ChplChapterResult() = default;
+
+  constexpr ChplChapterResult(ChplChapterStatus s) : status{s} {}
+
+  ChplChapterResult(ChplChapterStatus s, int64_t val) : status{s}, value{val} {}
+
+  ChplChapterResult(ChplChapterStatus s, std::string msg) : status{s}, errorMessage{std::move(msg)}
+  {
+  }
+
+  ChplChapterResult(ChplChapterStatus s, int64_t val, std::string msg)
+    : status{s}, value{val}, errorMessage{std::move(msg)}
+  {
+  }
+
+  explicit operator bool() const { return status == ChplChapterStatus::Found; }
+
+  bool isError() const { return status == ChplChapterStatus::Error; }
+  bool isFound() const { return status == ChplChapterStatus::Found; }
+  bool isNone() const { return status == ChplChapterStatus::None; }
+};
+
+class CChplChapterReader
+{
+public:
+  /**
+ * @brief Parses MP4 chapter atoms (chpl) from Nero/Ahead encoded files
+ *
+ * MP4 Atom Structure:
+ * - All atoms start with 4-byte size (big-endian) + 4-byte tag
+ * - If size == 1, next 8 bytes contain 64-bit size (big-endian)
+ * - Chapter atoms are typically nested: moov -> udta -> chpl
+ *
+ * CHPL Atom Format:
+ * - 1 byte: version (0 or 1)
+ * - 3 bytes: flags (currently unused)
+ * - Version 0: 1 byte chapter count
+ * - Version 1: 4 bytes time_base (big-endian) + 2 bytes count (little-endian)
+ * - For each chapter: 7 bytes timestamp + 1 byte title length + title string + 1 byte null
+ *   terminator
+ *
+ * Notes:
+ * - Chapter timestamp is in milliseconds for version 0, microseconds for version 1 unless there is
+ *   a value in the timebase in which case it will override the default values
+ * - Chapter names are utf-8 null-terminated strings
+ * - Total payload size must match atom size - 8 (header bytes)
+ *
+ * \param url [in] url of the file to be scanned
+ * \param chapters [out] vector containing the names and start times of any chapters found
+ * \return 1 if no chapters found, 0 on success, -1 on error, optionally returns an error message
+   */
+
+  static ChplChapterResult ScanNeroChapters(const CURL& url, std::vector<ChplChapter>& chapters);
+  static ChplChapterResult ParseChpl(XFILE::CFile& file,
+                                     const int64_t chpl_size,
+                                     std::vector<ChplChapter>& out);
+
+private:
+  static ChplChapterResult ReadAtomHeader(XFILE::CFile& file, uint32_t& tag, int& hdr_len);
```

Comment:
Could this function also be moved to the cpp file as free helper function in an unnamed namespace?

Reply:
I do not understand that.  The struct is defined in the header, why should the function in the cpp not be able to access it?